crates/chain-config/src/config/codec/decoder.rs

```diff
@@ -0,0 +1,170 @@
+use std::{
+    fmt::Debug,
+    fs::File,
+    io::Read,
+    path::{
+        Path,
+        PathBuf,
+    },
+};
+
+use super::parquet;
+use itertools::Itertools;
+
+use crate::{
+    config::{
+        contract_balance::ContractBalance,
+        contract_state::ContractStateConfig,
+    },
+    CoinConfig,
+    ContractConfig,
+    Group,
+    MessageConfig,
+    StateConfig,
+};
+
+use super::GroupResult;
+
+pub enum IntoIter<T> {
+    InMemory {
+        groups: std::vec::IntoIter<GroupResult<T>>,
+    },
+    Parquet {
+        decoder: parquet::Decoder<File, T>,
+    },
+}
+
+impl<T> Iterator for IntoIter<T>
+where
+    parquet::Decoder<File, T>: Iterator<Item = GroupResult<T>>,
+{
+    type Item = GroupResult<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            IntoIter::InMemory { groups } => groups.next(),
+            IntoIter::Parquet { decoder } => decoder.next(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum DataSource {
+    Parquet {
+        snapshot_dir: PathBuf,
+    },
+    InMemory {
+        state: StateConfig,
+        group_size: usize,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub struct Decoder {
+    data_source: DataSource,
+}
+
+impl Decoder {
+    pub fn json(
+        snapshot_dir: impl AsRef<Path>,
+        group_size: usize,
+    ) -> anyhow::Result<Self> {
+        let path = snapshot_dir.as_ref().join("state.json");
+
+        // This is a workaround until the Deserialize implementation is fixed to not require a
+        // borrowed string over in fuel-vm.
+        let mut contents = String::new();
+        let mut file = std::fs::File::open(path)?;
+        file.read_to_string(&mut contents)?;
+
+        let state = serde_json::from_str(&contents)?;
+
+        Ok(Self::in_memory(state, group_size))
+    }
+
+    pub fn in_memory(state: StateConfig, group_size: usize) -> Self {
+        Self {
+            data_source: DataSource::InMemory { state, group_size },
+        }
+    }
+
+    pub fn parquet(snapshot_dir: impl Into<PathBuf>) -> Self {
+        Self {
+            data_source: DataSource::Parquet {
+                snapshot_dir: snapshot_dir.into(),
+            },
+        }
+    }
+
+    pub fn detect_encoding(
+        snapshot_dir: impl AsRef<Path>,
+        default_group_size: usize,
+    ) -> anyhow::Result<Self> {
+        let snapshot_dir = snapshot_dir.as_ref();
+
+        if snapshot_dir.join("state.json").exists() {
```

Comment:
Hmm, it is not an ideal way of doing that=) Do we need this method?

Reply:
We can either ask the user to specify the encoding while reading the snapshot or we detect it for them. This is the latter approach.

I opted for detecting since a snapshot is now a directory and we're enforcing naming rules for the files within. If we have those expectations then detection can be robust.

Otherwise the CLI will get messy:

```
core run --regenesis --json dir/state.json
```
```
core run --regenesis --parquet --coins dir/coins.parquet --messages dir/messages.parquet ....
```

Reply:
I'll go with the metadata file route.